Makefile

```diff
@@ -78,48 +114,95 @@ version:
 	@echo $(VERSION)
 
 ./bin/log-counter: $(PKG_SOURCES)
-	CGO_ENABLED=$(CGO_ENABLED) GOOS=linux go build -o bin/log-counter \
+	CGO_ENABLED=$(CGO_ENABLED) GOOS=linux GOARCH=$(ARCH) go build -o bin/log-counter-$(ARCH) \
 	     -ldflags '-X $(PKG)/pkg/version.version=$(VERSION)' \
 	     $(BUILD_TAGS) cmd/logcounter/log_counter.go
 
 ./bin/node-problem-detector: $(PKG_SOURCES)
-	CGO_ENABLED=$(CGO_ENABLED) GOOS=linux go build -o bin/node-problem-detector \
+	CGO_ENABLED=$(CGO_ENABLED) GOOS=linux GOARCH=$(ARCH) go build -o bin/node-problem-detector-$(ARCH) \
 	     -ldflags '-X $(PKG)/pkg/version.version=$(VERSION)' \
 	     $(BUILD_TAGS) cmd/node_problem_detector.go
 
+qemu-register:
+	docker run --rm --privileged multiarch/qemu-user-static:register --reset
+
 Dockerfile: Dockerfile.in
 	sed -e 's|@BASEIMAGE@|$(BASEIMAGE)|g' $< >$@
+	sed -i "s|-ARCH|-$(QEMUARCH)|g" $@
+	sed -i "s|-BINARCH|-$(ARCH)|g" $@
+ifeq ($(ARCH),amd64)
+	# When building "normally" for amd64, remove the whole line, it has no part in the amd64 image
+	sed -i "/CROSS_BUILD_/d" Dockerfile
+else
+	curl -sSL https://github.com/multiarch/qemu-user-static/releases/download/$(QEMUVERSION)/x86_64_qemu-$(QEMUARCH)-static.tar.gz | tar -xz -C .
+	sed -i "s/CROSS_BUILD_//g" Dockerfile
+endif
 
 test: vet fmt
 	go test -timeout=1m -v -race ./cmd/options ./pkg/... $(BUILD_TAGS)
 
 build-binaries: ./bin/node-problem-detector ./bin/log-counter
 
-build-container: build-binaries Dockerfile
-	docker build -t $(IMAGE) .
+build-container-all: qemu-register $(addprefix build-container-,$(ALL_ARCH))
+
+build-container-%:
+	$(MAKE) ARCH=$* build-container
+
+build-container: clean build-in-docker Dockerfile
+	docker build -t $(IMAGE_WITH_ARCH) .
 
-build-tar: ./bin/node-problem-detector ./bin/log-counter
+build-tar: build-in-docker
```

Comment:
Why do we build tar in docker? I thought golang support multi-arch anyway?

Reply:
generic golang you can compile multi-arch but as we are using some of the cgo code, which requires some of the cross-tool libs which are present in `gcr.io/google_containers/kube-cross` so we have to build in the container.

Reply:
I talked to @Random-Liu offline and I think he preferred keeping the old way of building binaries. So adding him here to decide.